JoinColony/colonyCDapp
# Feat/#3189 advanced payment notifications

## Description

- Send notifications when an advanced payment is created and updated.

These are the notifications which should be sent:
- Ready for review
    - Trigger: When an advanced payment is created. This also sends a mention to any recipients.
- Ready for funding
    - Trigger: When an advanced payment has moved to the locked state and is ready for funding.
- Ready for release
    - Trigger: When an advanced payment is has been fully funded and can be released.
- Cancelled
    - Trigger: When an advanced payment has been cancelled.
- Finalized
    - Trigger: When an advanced payment payment has been finalized.

There is a block-ingestor side to this PR too, so please also check that out: https://github.com/JoinColony/block-ingestor/pull/273

## Testing

Ensure that the following is added to your `.env.local.secrets` and run `npm run prepare`.

```
MAGICBELL_API_KEY=<ask me for this>
MAGICBELL_API_SECRET=<ask me for this>
MAGICBELL_DEV_KEY=
```

* Log in as Leela
* Create a simple payment (just to check that notifications here still work) with Leela as the recipient, and check that you receive two notifications, one about the action, and one mention.
<img width="1070" alt="Screenshot 2024-10-02 at 12 18 00" src="https://github.com/user-attachments/assets/495c608c-c32b-4769-a9b1-e30cc7b44f62">
* Create an advanced payment, and ensure that Leela is one of the recipients, and check that you receive two notifications, one about the advanced payment being ready for review, and one mention.
<img width="1070" alt="Screenshot 2024-10-02 at 12 19 07" src="https://github.com/user-attachments/assets/f0e178fd-9f28-4dec-81b6-e5c6c5dc58fd">
<img width="676" alt="Screenshot 2024-10-02 at 12 19 22" src="https://github.com/user-attachments/assets/e7fca83c-6005-493c-b10e-e37b155f7780">
* Click on confirm details in the advanced payment's action sidepanel, and check that you receive a notification stating that the advanced payment is ready for funding.
<img width="678" alt="Screenshot 2024-10-02 at 12 20 00" src="https://github.com/user-attachments/assets/8c1281eb-161b-44fd-963c-3fe9ac7e930b">
* Fully fund the payment, and check that you receive a notification stating that the payment is ready for release.
<img width="1070" alt="Screenshot 2024-10-02 at 12 20 21" src="https://github.com/user-attachments/assets/5baae617-ad8f-48ed-8a15-32dbe0295b18">
<img width="673" alt="Screenshot 2024-10-02 at 12 20 35" src="https://github.com/user-attachments/assets/fe543002-2a5b-47f6-97d6-74c012e77166">
* Release the payment, and check that you receive a notification stating that the payment has been finalized.
<img width="1071" alt="Screenshot 2024-10-02 at 12 21 00" src="https://github.com/user-attachments/assets/ed52dcf9-ad7f-4f63-a837-79a8a2a6aabf">
<img width="672" alt="Screenshot 2024-10-02 at 12 21 14" src="https://github.com/user-attachments/assets/9f0e2d90-04fb-44b3-bdc1-2f862b651baa">
* Make a new advanced payment, this one without Leela as a recipient. Check that Leela only receives one notification this time (ie. no mention)
<img width="1069" alt="Screenshot 2024-10-02 at 12 22 13" src="https://github.com/user-attachments/assets/4aee5780-40bd-4555-8254-488c80dc8d8d">
* Cancel the payment, and check that you receive a notification about this:
<img width="864" alt="Screenshot 2024-10-02 at 12 22 46" src="https://github.com/user-attachments/assets/33f0a874-a947-4e93-86cf-32511214e3ae">
<img width="668" alt="Screenshot 2024-10-02 at 12 23 12" src="https://github.com/user-attachments/assets/46b49cbe-5b48-4738-b36f-c058b22112b4">


## Diffs

**New stuff** ✨

* New component `ExpenditureNotificationMessage`
* New properties added to NotificationType enum

**Changes** 🏗

* Changes to the block-ingestor: https://github.com/JoinColony/block-ingestor/pull/273
* Changes to the way notification messages are computed and displayed

Resolves #3189
